Using ObjectParser in MatchAllQueryBuilder and IdsQueryBuilder

cbuescher opened the pull request:
A first step moving away from the current parsing to use the generalized ObjectParser and ConstructingObjectParser. This PR start by making use of it in MatchAllQueryBuilder and IdsQueryBuilder.

cbuescher committed:
Using ObjectParser in MatchAllQueryBuilder and IdsQueryBuilder

A first step moving away from the current parsing to use the generalized
Objectparser and ConstructingObjectParser. This PR start by making use of it in
MatchAllQueryBuilder and IdsQueryBuilder.

nik9000 commented on core/src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -126,71 +119,20 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+        PARSER.declareStringArray((builder, values) -> builder.addIds(values.toArray(new String[values.size()])),
+                IdsQueryBuilder.VALUES_FIELD);
+        PARSER.declareFloat((builder, value) -> builder.boost(value), AbstractQueryBuilder.BOOST_FIELD);
+        PARSER.declareString((builder, value) -> builder.queryName(value), AbstractQueryBuilder.NAME_FIELD);
```
We might should move these last two declarations to a common spot something like
```
static <T extends AbstractObjectParser<? extends QueryBuilder>> declareStandardFields(T parser) {
        parser.declareFloat((builder, value) -> builder.boost(value), AbstractQueryBuilder.BOOST_FIELD);
        parser.declareString((builder, value) -> builder.queryName(value), AbstractQueryBuilder.NAME_FIELD);
        return parser;
}
```

and then we can declare them when we're initializing the object.

nik9000 replied:
I'm fairly sure I have the wrong generics incantation there....

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -308,7 +308,7 @@ public void testUnknownField() throws IOException {
         testQuery.queryName(marker); // to find root query to add additional bogus field there
         String queryAsString = testQuery.toString().replace("\"" + marker + "\"", "\"" + marker + "\", \"bogusField\" : \"someValue\"");
-        ParsingException e = expectThrows(ParsingException.class, () -> parseQuery(queryAsString));
+        RuntimeException e = expectThrows(RuntimeException.class, () -> parseQuery(queryAsString));
```
Maybe leave a comment that this is here for the duration of the migration or something. Could you add an assertion that the HTTP status code is still 400?

cbuescher replied:
That's a good point. Now I think ParsingExceptions would be better here. I think we should wrap the IllegalArgumentException that the object parsers throw currently in a ParsingException in fromXContent() then.

nik9000 commented on core/src/test/java/org/elasticsearch/index/query/IdsQueryBuilderTests.java:
```diff
@@ -144,8 +155,8 @@ public void testFromJsonDeprecatedSyntax() throws IOException {
-                "        \"types\" : [\"" + type + "\"],\n" +
-                "        \"values\" : []\n" +
+                "        \"types\" : [\"my_type\"],\n" +
+                "        \"values\" : [\"1\"]\n" +
```
Is it not ok to have an empty array anymore? That seems weird actually.

nik9000 commented on core/src/test/java/org/elasticsearch/index/query/IdsQueryBuilderTests.java:
```diff
@@ -32,17 +32,8 @@
-        String noIdsFieldQuery = "{\"ids\" : { \"type\" : \"my_type\"  }";
-        ParsingException e = expectThrows(ParsingException.class, () -> parseQuery(noIdsFieldQuery));
-        assertThat(e.getMessage(), containsString("no ids values provided"));
-    }
```
We should still throw an exception here though.

nik9000 replied:
OK!

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -126,71 +119,20 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
-        }
+    @SuppressWarnings("unchecked")
+    private static ConstructingObjectParser<IdsQueryBuilder, QueryParseContext> PARSER = new ConstructingObjectParser<>(NAME,
+            a -> new IdsQueryBuilder(((List<String>) a[0]).toArray(new String[((List<String>) a[0]).size()])));
```
Great, I switched to empty array.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -126,71 +119,20 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
-        return Optional.of(query);
+    static {
+        PARSER.declareStringArray(constructorArg(), IdsQueryBuilder.TYPE_FIELD);
+        PARSER.declareStringArray((builder, values) -> builder.addIds(values.toArray(new String[values.size()])),
```
It seem to do, I changed a test so that it checks that parsing numbers works.

cbuescher commented:
@nik9000 thanks for your input, I pushed an update adressing most of your comments. I think not throwing errors when `values` is missing is okay because it will still produce a valid query.

cbuescher committed:
Don't require `types` parameter in IdsQueryBuilder constructor

According to the docs and our own tests we accept an ids query without specified
types and default to all types in the index mapping in this case. This changes
the builder to reflect this by making the types no longer a required constructor
argument and changes the parser to reflect that.

cbuescher commented:
@nik9000 I added another update after a failing test in CI that concerns the optional `types` parameter, could you take another look?

nik9000 reviewed (approved):
Sorry it took me so long to notice this! I'd forgotten about it.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java:
```diff
@@ -300,4 +302,12 @@ protected static void throwParsingExceptionOnMultipleFields(String queryName, XC
+     * Adds 'boost' and 'query_name' parsing to all query builder parsers passed in
+     */
+    protected static void declareStandardFields(AbstractObjectParser<? extends QueryBuilder, ? extends ParseFieldMatcherSupplier> parser) {
+        parser.declareFloat((builder, value) -> builder.boost(value), AbstractQueryBuilder.BOOST_FIELD);
```
You mean `AbstractQueryBuilder::boost` I tried this for a while now but can't get it to work, there's something going on with the generics in AbstractQueryBuilder that seems to prevent this. I'll leave it for now as it seems to work.

cbuescher commented:
Also on 5.x: 649b6c7c6007a8340850b905e82a4335d4fa2748 and 6ab528c7d04c734346f5afcf8a38737fab99952c
